### DISCUSSION TITLE
Fastapi + pudo/dataset

### DISCUSSION THREAD

**Discussion by remidinishanth-ntnx** [fastapi/fastapi#7286] [Questions]
I'm planning to use [dataset](https://github.com/pudo/dataset) to connect to the Postgres database(I wanted to store huge XML data and without writing a SQL model schema). Are there any downsides of using dataset with fastapi? Also, where can I do the db.connect(uri)? Can this be done just one globally or should I call it once for every API call?

I've seen https://github.com/tiangolo/fastapi/issues/726.

Dataset directly creates engine https://github.com/pudo/dataset/blob/master/dataset/database.py#L56 and uses `engine.connect()` instead of `sessionmaker`. Since connect creates engine I feel it's not good to call it for every API call, so how I can I make it global and 

Can you help me with where the DB connect has to go and what works best with fastapi? Is there any way I can use it as a dependency?

**Comment by raphaelauv** [comment]
You manage the connection as you want/need , if it's one connection for every call then you can create it inside on_startup and store the connection inside a singleton.

Well your question is a backend generic question and there is NOTHING specific to fastAPI.
You would have the same question with ANY web-framework.

If you agree just close this issue

**Comment by remidinishanth-ntnx** [comment]
My question was more around these lines

```
SQLAlchemy scoped_session is based on thread locals. https://github.com/tiangolo/fastapi/issues/726#issuecomment-584371305
```

If I use a global connection using `@app.on_event("startup")` will there be any issues

**Comment by raphaelauv** [comment]
This zone is  `issue for fastAPI` not stackoverflow , if you do not have a specific and reproducible problem with fastAPI then close this issue

**Comment by tiangolo** [comment]
Thanks for the help here @raphaelauv ! :clap: :bow: 

Thanks for reporting back and closing the issue @remidinishanth-ntnx :+1:

I have no idea about that library, but it seems it handles its sessions independently using thread locals: https://github.com/pudo/dataset/blob/master/dataset/database.py#L46

Those thread locals are not compatible with async/await. They should be replaced by contextvars that are compatible with async/await (as well as threads). An example fo the problem and how to patch it in your code (if you want to do that) is here: https://fastapi.tiangolo.com/advanced/sql-databases-peewee/
